ENH: interpolate: Array API for PPoly and related objects

ev-br opened the pull request:
#### Reference issue
<!--Example: Closes gh-WXYZ.-->

Continues gh-23052 ; cross-ref https://github.com/scipy/scipy/issues/23754 

#### What does this implement/fix?
<!--Please explain your changes.-->

Convert

- [x] PPoly
- [x] BPoly
- [x] CubicHermiteSpline
- [x] PchipInterpolator
- [x] Akima1DInterpolator
- [x] CubicSpline    
.
- [ ] PPoly.roots
- [ ] BPoly.from_derivatives


#### Additional information
<!--Any additional information you think is important.-->

Two not-entirely-trivial items, both related to (the lack of) ragged arrays:

- `PPoly.roots` returns either a numeric array or an object array, depending on the data. We discussed it a while ago, https://github.com/scipy/scipy/pull/3260#issuecomment-1166510296, and the conclusion was, basically, "yes it's ugly but not worth breaking backwards compat for". That was before array API though. The only Array API compatible way I can imagine is to make it return either an array or a list of arrays. While not great, it is similar in spirit to what it does now.
- `BPoly.from_derivatives` accepts either an array or a list of arrays. The latter we cannot avoid because otherwise we'd need ragged arrays.

ev-br committed:
ENH: interpolate: array API PPoly and BPoly

ev-br committed:
ENH: interpolate: Array API for _cubic interpolators

CubicHermiteSpline, PCHIP, Akima1D, CubicSpline

ev-br committed:
TST: interpolate: add CPU-only skips

ev-br committed:
ENH: interpolate: make CubicHermiteSpline work with JAX immutable arrays

ev-br committed:
BUG: interpolate: do not mix numpy and other arrays

ev-br committed:
TST: interpolate: xfail Akima1D tests on Jax

ev-br committed:
TST: interpolate: xfail tests on Dask

ev-br committed:
TST: interpolate: fix testing with f32 default

ev-br commented:
Rebased on main to remove things merged in gh-23052. This is now ready for review. The missing "weird/specialist" parts, `PPoly.roots` and `BPoly.from_derivatives` can be handled in a follow-up, I believe.

j-bowhay commented on scipy/interpolate/_cubic.py:
```diff
@@ -138,20 +142,23 @@ class CubicHermiteSpline(PPoly):
         if extrapolate is None:
             extrapolate = True
 
         x, dx, y, axis, dydx = prepare_input(x, y, axis, dydx)
```
```suggestion
        x, dx, y, axis, dydx = prepare_input(x, y, axis, dydx, xp=xp)
```

j-bowhay commented on scipy/interpolate/_cubic.py:
```diff
@@ -13,34 +15,36 @@
     axis. The value of `axis` is converted to lie in
     [0, number of dimensions of `y`).
     """
+    if xp is None:
+        xp = array_namespace(x, y, dydx)
```
I would prefer if we didn't do this as it would help root out any instances we don't pass `xp`
```suggestion
```

j-bowhay commented on scipy/interpolate/_cubic.py:
```diff
@@ -556,7 +566,9 @@ def __init__(self, x, y, axis=0, *, method: Literal["akima", "makima"]="akima",
 
             # These are the mask of where the slope at breakpoint is defined:
-            ind = np.nonzero(f12 > break_mult * np.max(f12, initial=-np.inf))
+            mmax = xp.asarray(np.max(np.asarray(f12), initial=-np.inf))
+            ind = xp.nonzero(f12 > break_mult * mmax)
```
Possibly?
```suggestion
            mmax = xp.max(f12) if f12.size > 0 else -xp.inf
            ind = xp.nonzero(f12 > break_mult * mmax)
```

ev-br replied:
Thanks, done. This `initial=` thingy is not a little esoteric indeed.

j-bowhay commented on scipy/interpolate/tests/test_interpolate.py:
```diff
@@ -833,52 +836,62 @@ def test_lagrange(self):
         assert_array_almost_equal(p.coeffs,pl.coeffs)
 
 
+@xfail_xp_backends("dask.array", reason="lacks nd fancy indexing")
+@xfail_xp_backends("jax.numpy", reason="immutable arrays")
+@xfail_xp_backends("array_api_strict", reason="fancy indexing __setitem__")
+@skip_xp_backends(cpu_only=True)
```
It would be quite nice to use the `xp_capabilities` / `make_xp_skip` machinery now it's quite well established

ev-br replied:
This is addressed now, thanks Jake!

ev-br committed:
MAINT: interpolate: Apply suggestions from code review

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

j-bowhay committed:
ENH: interpolate: use xp_capabilities

j-bowhay reviewed (commented):
I pushed a commit to try using `xp_capabilities`, it also looked like there were a fair number of tests that are still to be converted?

j-bowhay commented on scipy/interpolate/_cubic.py:
```diff
@@ -50,25 +52,26 @@ def prepare_input(x, y, axis, dydx=None):
     return x, dx, y, axis, dydx
 
 
+@xp_capabilities(cpu_only=True)
```
```suggestion
@xp_capabilities(cpu_only=True, jax_jit=False)
```

j-bowhay commented on scipy/interpolate/_cubic.py:
```diff
@@ -378,6 +385,11 @@ def pchip_interpolate(xi, yi, x, der=0, axis=0):
+@xp_capabilities(cpu_only=True, xfail_backends=[
+    ("dask.array", "lacks nd fancy indexing"),
+    ("jax.numpy", "immutable arrays"),
+    ("array_api_strict", "fancy indexing __setitem__")
```
```suggestion
    ("array_api_strict", "fancy indexing __setitem__"),
    jax_jit=False
```

j-bowhay commented on scipy/interpolate/_cubic.py:
```diff
@@ -584,6 +600,7 @@ def from_bernstein_basis(cls, bp, extrapolate=None):
                                   "an Akima interpolator.")
 
 
+@xp_capabilities(cpu_only=True)
```
```suggestion
@xp_capabilities(cpu_only=True, jax_jit=False)
```

j-bowhay commented on scipy/interpolate/_interpolate.py:
```diff
@@ -802,21 +823,22 @@ def __call__(self, x, nu=0, extrapolate=None):
+        return self._asarray(out)
 
 
+@xp_capabilities(cpu_only=True)
```
```suggestion
@xp_capabilities(cpu_only=True, jax_jit=False)
```

j-bowhay commented on scipy/interpolate/_interpolate.py:
```diff
@@ -1262,24 +1288,25 @@ def from_bernstein_basis(cls, bp, extrapolate=None):
+        return cls.construct_fast(bp._asarray(c), bp.x, extrapolate, bp.axis)
 
 
+@xp_capabilities(cpu_only=True)
```
```suggestion
@xp_capabilities(cpu_only=True, jax_jit=False)
```

j-bowhay commented on scipy/interpolate/_cubic.py:
```diff
@@ -378,6 +385,12 @@ def pchip_interpolate(xi, yi, x, der=0, axis=0):
+    ("dask.array", "lacks nd fancy indexing"),
+    ("jax.numpy", "immutable arrays"),
+    ("array_api_strict", "fancy indexing __setitem__"),
+    jax_jit=False
```
```suggestion
```

j-bowhay commented on scipy/interpolate/_interpolate.py:
```diff
@@ -802,21 +823,22 @@ def __call__(self, x, nu=0, extrapolate=None):
+        return self._asarray(out)
 
 
+@xp_capabilities(cpu_only=True, jax_jit=False)
```
```suggestion
@xp_capabilities(cpu_only=True, jax_jit=False, allow_dask_compute=1)
```

j-bowhay commented on scipy/interpolate/_cubic.py:
```diff
@@ -50,25 +52,26 @@ def prepare_input(x, y, axis, dydx=None):
     return x, dx, y, axis, dydx
 
 
+@xp_capabilities(cpu_only=True, jax_jit=False)
```
```suggestion
@xp_capabilities(cpu_only=True, jax_jit=False, allow_dask_compute=1)
```

ev-br reviewed (commented):
Two minor annoyances from `xp_capabilities`, on top of CI complaints which seem to be related.

steppi commented:
I investigated the problem with `PPoly` on dask and found that it's due to [make_xp_pytest_marks](https://github.com/scipy/scipy/blob/71fb82e33c47a4358804578a6a16ed9129ef4049/scipy/_lib/_array_api.py#L896) calling `array_api_extra`'s function `lazy_xp_function` on tested things, which on dask wraps them with [_dask_wrap](https://github.com/data-apis/array-api-extra/blob/20593693e22bcfc0fddee27fcacc91abc8c8dbf7/src/array_api_extra/testing.py#L382). When using `make_xp_pytest_marks` (which `make_xp_test_case` and others use under the hood) with functions, this is fine, but it means that when using `make_xp_pytest_marks` with classes, your class gets replaced with a function which only implements the `__call__` method. I think the ideal solution would be for `array_api_extra` to allow support for something like `lazy_xp_function` for classes, so I raised an issue here https://github.com/data-apis/array-api-extra/issues/488 asking for an enhancement request. @ev-br, in the meantime I think it's fine to not test classes with `lazy_xp_function` at all. At the moment there seems to be no way to avoid it though. Something needs to change in either `scipy._lib._array_api` or in array_api_extra.

ev-br committed:
MAINT: interpolate: skip testing PPoly et al on dask

xp-capabilities is broken with classes, and there is not
much help from dask anyway for interpolate classes.

ev-br commented:
Thanks @steppi. I updated the PR to skip testing on dask until https://github.com/data-apis/array-api-extra/issues/488 is fixed one way or another. 
While at it, I squashed @j-bowhay 's commits into a single commit, so that the PR is mergeable.
CI failures are unrelated.

ev-br commented on scipy/interpolate/_cubic.py:
```diff
@@ -50,25 +52,26 @@ def prepare_input(x, y, axis, dydx=None):
     return x, dx, y, axis, dydx
 
 
+@xp_capabilities(cpu_only=True, jax_jit=False, allow_dask_compute=1)
```
Dask-related polymorphism  is tracked at https://github.com/data-apis/array-api-extra/issues/415. While I still think this is a wart, it should not be blocking here.

ev-br commented on scipy/interpolate/_cubic.py:
```diff
@@ -584,6 +600,7 @@ def from_bernstein_basis(cls, bp, extrapolate=None):
                                   "an Akima interpolator.")
 
 
+@xp_capabilities(cpu_only=True, jax_jit=False, allow_dask_compute=1)
```
The immediate issue was fixed. Fiddling with the class docstring as opposed to the instance docstring is unfortunate, but is not blocking ilthis PR. Resolving.

steppi reviewed (approved):
Looks good to me. The failures are indeed unrelated. Thanks @ev-br!
